### PR TITLE
refactor(gui): Refactor file picker fs I/O

### PR DESCRIPTION
**Pull request by jhermsmeier**
This refactors the experimental file picker to avoid fs i/o in as many places as possible to improve performance. Further, rendering performance is improved by avoiding unnecessary element state changes invalidating components. Also, recent files & favorites have been temporarily disabled due to lack of need for Etcher Pro.

screen shot 2018-06-11 at 19 49 23

screen shot 2018-06-11 at 19 50 07

Errors / warnings from within the file picker are native OS dialogs, as this is a React component and doesn't have access to the Angular modal controllers / components;

screen shot 2018-06-11 at 19 50 15

screen shot 2018-06-11 at 19 59 06

Change-Type: patch
Connects To: https://github.com/resin-io/etcher/issues/2348, https://github.com/resin-io/etcher/issues/1823

### CODE REVIEW

File: lib/gui/app/components/file-selector/file-selector/file-list.jsx [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+const supportedFormats = require('../../../../../shared/supported-formats')
+
+const debug = require('debug')('etcher:gui:file-selector')
+
+/**
+ * @summary Character limit of a filename before a middle-ellipsis is added
+ * @constant
+ * @private
+ */
+const FILENAME_CHAR_LIMIT = 20
```
Comment (zvin) [inline lib/gui/app/components/file-selector/file-selector/file-list.jsx]: This is not used anywhere.

File: lib/gui/app/components/file-selector/file-selector/file-list.jsx [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+  componentDidMount () {
+    process.nextTick(() => {
+      this.readdir(this.state.path)
+    })
+  }
+
+  onHighlight (file) {
+    debug('FileList:onHighlight', file)
+    // this.setState({ highlighted: file.path })
```
Comment (zvin) [inline lib/gui/app/components/file-selector/file-selector/file-list.jsx]: Shouold this be there?

File: lib/gui/app/components/file-selector/file-selector/file-list.jsx [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    this.props.onHighlight(file)
+  }
+
+  onSelect (file) {
+    debug('FileList:onSelect', file.path, file.isDirectory)
+    this.props.onSelect(file)
+  }
+
+  componentDidUpdate () {
+    console.timeEnd('FileList:componentDidUpdate')
```
Comment (zvin) [inline lib/gui/app/components/file-selector/file-selector/file-list.jsx]: Shouldn't we use `debug` for that?

File: lib/gui/app/models/files.js [hunk 4]
```diff
@@ -16,133 +16,92 @@
...
+    this.dirname = components.dir
+    this.basename = components.base
+    this.name = components.name
+    this.ext = components.ext
+    this.isHidden = components.name.startsWith('.')
+    this.isFile = stats.isFile()
+    this.isDirectory = stats.isDirectory()
+    this.isBlockDevice = stats.isBlockDevice()
+    this.isCharacterDevice = stats.isCharacterDevice()
+    this.stats = stats
```
Comment (zvin) [inline lib/gui/app/models/files.js]: the only thing left in `stats` that we use is `size`. Why not copy it like the rest and remove `FileEntry.stats`?

**Comment by lurch** [comment]
Looking at those screenshots - how does this file-picker work if you have two different images which get shortened to the same ellipses-filename? (which I guess is quite likely if you e.g. have different versions of Raspbian in the same folder?)

**Comment by jhermsmeier** [comment]
@lurch yeah, it doesn't work very well in that situation, which is something we've already noticed; will need some more design / ux work – but then, this is experimental stuff, so it'll be rough round the edges atm

File: package.json [hunk 5]
```diff
@@ -40,8 +40,7 @@
     "fsevents"
   ],
   "dependencies": {
-    "@fortawesome/fontawesome": "1.1.5",
-    "@fortawesome/fontawesome-free-solid": "5.0.9",
+    "@fortawesome/fontawesome-free-webfonts": "^1.0.9",
```
Comment (curcuz) [inline package.json]: @jhermsmeier does this mean we depend on internet connection now?

Comment (jhermsmeier) [reply package.json]: Nope, the font itself is in the npm package

**Comment by lurch** [comment]
> it doesn't work very well in that situation, which is something we've already noticed

I guess one option would be to display the full filename on mouse-hover? Although I guess that doesn't help if you're using a touch screen ;)

File: lib/gui/app/components/file-selector/file-selector/file-list.jsx (hunk 3 above)
Comment (jhermsmeier) [reply lib/gui/app/components/file-selector/file-selector/file-list.jsx]: `debug` doesn't support timing like this that mark timings in the devtools' perf timeline

File: lib/gui/app/models/files.js (hunk 4 above)
Comment (jhermsmeier) [reply lib/gui/app/models/files.js]: Yeah true, done

File: lib/gui/app/components/file-selector/file-selector/file-list.jsx (hunk 2 above)
Comment (jhermsmeier) [reply lib/gui/app/components/file-selector/file-selector/file-list.jsx]: Nope, removed

File: lib/gui/app/components/file-selector/file-selector/file-list.jsx (hunk 1 above)
Comment (jhermsmeier) [reply lib/gui/app/components/file-selector/file-selector/file-list.jsx]: Good catch, fixed!

**Comment by jhermsmeier** [comment]
@konmouz wdyt about this updated fade color:

screen shot 2018-06-12 at 18 51 41

**Pull request merged**